### PR TITLE
Fix issue 1875 learning curve keyerror with training only statistics

### LINKED ISSUE #1875: learning_curves plot fails with KeyError with training only statistics generated
**Describe the bug**
if `training_statistics.json` contains only training results and no validation results and is used to as input to the `learning_curves` plot, the plot fails with `KeyError` exception.   Here is stack trace:
```
Traceback (most recent call last):
  File "/usr/local/lib/python3.8/site-packages/IPython/core/interactiveshell.py", line 3369, in run_code
    exec(code_obj, self.user_global_ns, self.user_ns)
  File "<ipython-input-2-a90fbbbbb49e>", line 1, in <cell line: 1>
    runfile('/opt/project/sandbox/mwe_learning_curve_viz_failure.py', wdir='/opt/project/sandbox')
  File "/opt/.pycharm_helpers/pydev/_pydev_bundle/pydev_umd.py", line 198, in runfile
    pydev_imports.execfile(filename, global_vars, local_vars)  # execute the script
  File "/opt/.pycharm_helpers/pydev/_pydev_imps/_pydev_execfile.py", line 18, in execfile
    exec(compile(contents+"\n", file, 'exec'), glob, loc)
  File "/opt/project/sandbox/mwe_learning_curve_viz_failure.py", line 54, in <module>
    learning_curves([training_stats2],
  File "/opt/project/ludwig/visualize.py", line 1245, in learning_curves
    validation_stats.append(learning_stats[VALIDATION][output_feature_name][metric])
KeyError: 'combined'
```

I'll submit a PR to address this issue later tonight.

**To Reproduce**
This is a minimal working example demonstrating the issue:
```
import logging
import os
import tempfile
import shutil
import json

from ludwig.api import LudwigModel
from tests.integration_tests.utils import generate_data, number_feature
from ludwig.visualize import learning_curves


if __name__ == "__main__":
    input_features = [number_feature()]
    output_features = [number_feature()]
    config = {
        "input_features": input_features,
        "output_features": output_features,
        "trainer": {"epochs": 5}
    }

    with tempfile.TemporaryDirectory() as tmpdir:
        data_fp = generate_data(input_features, output_features, os.path.join(tmpdir, "raw.csv"))

        model = LudwigModel(config=config)

        # using dataset=data_fp the both calls to generate learning_curves work
        #    which generates training and validation statistics
        # using training_set=data_fp will cause the second learning_curves plot to fail
        #    which generates training only statistics
        training_stats1, _, results = model.train(
           # dataset=data_fp,   # this works in all cases
            training_set=data_fp, # this will cause a failure
            output_directory=os.path.join(tmpdir, 'results')
        )

        with open(os.path.join(results, "training_statistics.json")) as f:
            training_stats2 = json.load(f)

        print(training_stats2)

        # training statistics data structure created by train() works for two situations:
        #    training only statistics captured
        #    training and validation statistics generated
        shutil.rmtree('./viz1', ignore_errors=True)
        learning_curves([training_stats1],
                        model_names=['Model1'],
                        output_directory="./viz1", file_format='png')
        print("generated plots for training_stats1", os.listdir("./viz1"))

        # using training stats dictionary loaded from training_statistics.json fails in one situation:
        #     training only statistics fails
        #     training and validation statistics works
        shutil.rmtree('./viz2', ignore_errors=True)
        learning_curves([training_stats2],
                        model_names=['Model1'],
                        output_directory="./viz2", file_format='png')
        print("generated plots for training_stat2", os.listdir("./viz2"))

    print("all done")
```

**Expected behavior**
Generate the learning curve plot with training only statistics

**Screenshots**
Full output from the example code:
```
NumExpr defaulting to 5 threads.
generated new fontManager
ray.init() failed: Could not find any running Ray instance. Please specify the one to connect to by setting `address`.
Note: steps_per_checkpoint (was 1) is now set to the number of steps per epoch: 1.
{'test': {}, 'training': {'combined': {'loss': [0.7418510913848877, 0.7391425967216492, 0.7364398241043091, 0.7337425351142883, 0.7310507893562317]}, 'num_BD40A': {'loss': [0.7418510913848877, 0.7391425967216492, 0.7364398241043091, 0.7337425351142883, 0.7310507893562317], 'mean_absolute_error': [0.8156701922416687, 0.8140162825584412, 0.8123624324798584, 0.8107086420059204, 0.8090547919273376], 'mean_squared_error': [0.7418510913848877, 0.7391425967216492, 0.7364398241043091, 0.7337425351142883, 0.7310507893562317], 'r2': [-8.04835319519043, -8.0153169631958, -7.982351303100586, -7.9494524002075195, -7.916621208190918], 'root_mean_squared_error': [0.8613077998161316, 0.8597339987754822, 0.8581607341766357, 0.8565877079963684, 0.8550150990486145], 'root_mean_squared_percentage_error': [3.0594561100006104, 3.045417070388794, 3.031384229660034, 3.0173561573028564, 3.003333806991577]}}, 'validation': {}}
generated plots for training_stats1 ['learning_curves_combined_loss.png', 'learning_curves_num_BD40A_loss.png']
Traceback (most recent call last):
  File "/usr/local/lib/python3.8/site-packages/IPython/core/interactiveshell.py", line 3369, in run_code
    exec(code_obj, self.user_global_ns, self.user_ns)
  File "<ipython-input-2-a90fbbbbb49e>", line 1, in <cell line: 1>
    runfile('/opt/project/sandbox/mwe_learning_curve_viz_failure.py', wdir='/opt/project/sandbox')
  File "/opt/.pycharm_helpers/pydev/_pydev_bundle/pydev_umd.py", line 198, in runfile
    pydev_imports.execfile(filename, global_vars, local_vars)  # execute the script
  File "/opt/.pycharm_helpers/pydev/_pydev_imps/_pydev_execfile.py", line 18, in execfile
    exec(compile(contents+"\n", file, 'exec'), glob, loc)
  File "/opt/project/sandbox/mwe_learning_curve_viz_failure.py", line 54, in <module>
    learning_curves([training_stats2],
  File "/opt/project/ludwig/visualize.py", line 1245, in learning_curves
    validation_stats.append(learning_stats[VALIDATION][output_feature_name][metric])
KeyError: 'combined'
Process finished with exit code 0
```

**Environment (please complete the following information):**

- OS: MacOS running Docker Desktop
- Version: MacOS 12.2.1 , Docker Desktop 4.6.0
- Python version: 3.8
- Ludwig version: 0.5rc2

**Additional context**
The failure only occurs using the python dictionary created from reading in `training_statistics.json` from disk. 

The error **does not occur** if using the trainining statistics data structure created by `LudwigModel.train()`

### PR DESCRIPTION
**Pull request by jimthompson5802**
# Code Pull Requests

Fix #1875 

Summary of changes:
* updated  `visualize.py` to handle use correctly detect when validation metrics are not present in training_statistics
* updated unit test to test the situation that caused issue #1875 

Output from updated unit test:
```
root@0b8156dfe700:/opt/project/sandbox/pytest_area# pytest -v /opt/project/tests/integration_tests/test_visualization_api.py::test_learning_curves_vis_api
=================================================================================== test session starts ====================================================================================
platform linux -- Python 3.8.13, pytest-7.1.1, pluggy-1.0.0 -- /usr/local/bin/python
cachedir: .pytest_cache
rootdir: /opt/project, configfile: pytest.ini
plugins: timeout-2.1.0, xdist-2.5.0, pycharm-0.7.0, forked-1.4.0, cov-3.0.0, anyio-3.5.0
collected 2 items

../../tests/integration_tests/test_visualization_api.py::test_learning_curves_vis_api[True] PASSED                                                                                   [ 50%]
../../tests/integration_tests/test_visualization_api.py::test_learning_curves_vis_api[False] PASSED                                                                                  [100%]

```

### CODE REVIEW

File: ludwig/visualize.py [hunk 1]
```diff
@@ -1241,7 +1241,7 @@ def learning_curves(
 
                 validation_stats = []
                 for learning_stats in train_stats_per_model_list:
-                    if VALIDATION in learning_stats:
+                    if VALIDATION in learning_stats and len(learning_stats[VALIDATION]) > 0:
```
Comment (dantreiman) [inline ludwig/visualize.py:1244]: I think checking for `output_feature_name in learning_stats[VALIDATION]`
```suggestion
                    if VALIDATION in learning_stats and output_feature_name in learning_stats[VALIDATION]:
```
is a bit more self-documenting.

**Review by dantreiman — APPROVED** [review]
Added one minor comment, overall looks good to me.  Thanks!

File: ludwig/visualize.py (hunk 1 above)
Comment (jimthompson5802) [reply ludwig/visualize.py]: @dantreiman   Thank you for the comment.  I incorporated your suggestion.

**Comment by w4nderlust** [comment]
Thank you @jimthompson5802 !

**Pull request merged**